File: C++/19.备忘录模式/MementoPatternDemo.cpp
/**
 * @file MementoPatternDemo.cpp
 * @author {Layne} ({shu_huanglei@163.com})
 * @brief
 * @version 0.1
 * @date 2019-11-20
 *
 * @copyright Copyright (c) 2019
 *
 */

#include "CareTaker.h"
#include "Originator.h"

int main(int argc, char const *argv[]) {
    std::shared_ptr<Originator> originator = std::make_shared<Originator>();
    std::shared_ptr<CareTaker> care_taker = std::make_shared<CareTaker>();
    originator->SetState("State #1");
    originator->SetState("State #2");
    care_taker->Add(originator->SaveStateToMemento());
    originator->SetState("State #3");
    care_taker->Add(originator->SaveStateToMemento());
    originator->SetState("State #4");

    std::cout << "Current State: " << originator->GetState() << std::endl;
    originator->getStateFromMemento(care_taker->Get(0));
    std::cout << "First saved State: " << originator->GetState() << std::endl;
    originator->getStateFromMemento(care_taker->Get(1));
    std::cout << "Second saved State: " << originator->GetState() << std::endl;

    return 0;
}

File: Objective-C/19.MementoPattern/MementoPatternDemo.m
//
//  MementoPatternDemo.m
//  19.MementoPattern
//
//  Created by Layne on 2020/11/12.
//

#import "CareTaker.h"
#import "Originator.h"

int main(int argc, const char* argv[]) {
    @autoreleasepool {
        Originator* originator = [[Originator alloc] init];
        CareTaker* careTaker = [[CareTaker alloc] init];

        [originator setState:@"State #1"];
        [originator setState:@"State #2"];
        [careTaker add:[originator saveStateToMemento]];
        [originator setState:@"State #3"];
        [careTaker add:[originator saveStateToMemento]];
        [originator setState:@"State #4"];

        NSLog(@"Current State: %@", [originator getState]);
        [originator getStateFromMemento:[careTaker get:0]];
        NSLog(@"First saved State: %@", [originator getState]);
        [originator getStateFromMemento:[careTaker get:1]];
        NSLog(@"Second saved State: %@", [originator getState]);
    }
    return 0;
}
